compiler/codegen/OMRTreeEvaluator.cpp

```diff
@@ -51,6 +51,418 @@
+   }
+
+TR::Register *
+OMR::TreeEvaluator::drdbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   // A drdbar IL opcode is essentially a dload IL opcode with side effects. We first handle
+   // the side effect, then call the appropriate load evaluator to properly evaluate the opcode.
+
+   // Currently drdbar nodes don't have any side effects and are effectively just a load operation. 
+   // However we must evaluate drdbar's child here since it is the responsibility of the evaluator
+   // to evaluate all of its children. Additionally, the dloadEvaluator is not aware of drdbar's child
+   // and will not use it. It is only needed for the side effect (i.e. by this evaluator).
+   TR::Node *firstChild = node->getFirstChild();
+   cg->evaluate(firstChild);
+   cg->decReferenceCount(firstChild);
+   return TR::TreeEvaluator::dloadEvaluator(node, cg);
+   }
+
+TR::Register *
+OMR::TreeEvaluator::drdbariEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   // A drdbari IL opcode is essentially a dloadi IL opcode with side effects. We first handle
+   // the side effect, then call the appropriate load evaluator to properly evaluate the opcode.
+
+   // Currently drdbari nodes don't have any side effects and are effectively just a load operation. 
+   // However we must evaluate drdbari's child here since it is the responsibility of the evaluator
+   // to evaluate all of its children.
+   cg->evaluate(node->getFirstChild());
+   // Note: We don't decrement the reference count because the dloadEvaluator
+   // expects the child and will decrement the reference count for us. Thus decrementing the reference
+   // count here would result in a double decrement (which is incorrect).
+
+   return TR::TreeEvaluator::dloadEvaluator(node, cg);
+   }
+
+TR::Register *
+OMR::TreeEvaluator::bwrtbariEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   // An bwrtbari IL opcode is essentially an bstorei IL opcode with side effects. We first handle
+   // the side effect, then call the appropriate store evaluator to properly evaluate the opcode.
+
+   // We evaluate the children here because it's the evaluator's responsibility to evaluate all of its
+   // children.
+   TR::Node *sideEffectNode = node->getThirdChild();
+   cg->evaluate(node->getFirstChild());
+   cg->evaluate(node->getSecondChild());
+   cg->evaluate(sideEffectNode);
+
+   // Note: We only decrement the reference count for the side effect node. This is because
+   // the store evaluator is not aware of wrtbar's side effect node and so will not evaluate/decrement the 
+   // reference count for it. However it is aware and will do this work for the other children. So we skip
+   // decrementing the reference count for the other children here and let the store evaluator handle it in
+   // order to avoid double decrementing the reference count.
+   cg->decReferenceCount(sideEffectNode);
+
+   return TR::TreeEvaluator::bstoreEvaluator(node, cg);
```

Comment:
The write barrier evaluators are too aggressive. They should only be evaluating nodes which they are handling. In this case the side effect node is the only one we should be evaluating. The delegated store evaluator will take care of evaluating the store portion of the write barrier. These evaluators should look like this:

```
TR::Register *
OMR::TreeEvaluator::bwrtbariEvaluator(TR::Node *node, TR::CodeGenerator *cg)
   {
   // A bwrtbari IL has a side effect node which is currently not used at the OMR level. We first handle the side
   // effect, then call the appropriate store evaluator to properly evaluate the opcode.
   TR::Node *sideEffectNode = node->getThirdChild();
   cg->evaluate(sideEffectNode);

   // Note: We only decrement the reference count for the side effect node. This is because the store evaluator is not
   // aware of wrtbar's side effect node and so will not evaluate/decrement the reference count for it. However it is
   // aware and will do this work for the other children. So we skip decrementing the reference count for the other
   // children here and let the store evaluator handle it in order to avoid double decrementing the reference count.
   cg->decReferenceCount(sideEffectNode);

   return TR::TreeEvaluator::bstoreEvaluator(node, cg);
   }
```

Reply:
Similarly for all other write barrier evaluators.